### PR TITLE
perf: Docker image improvements

### PR DESCRIPTION
**Pull request by AlisaAkiron**
### Description

Docker image improvements

For Dockerfile:
- Use `ENTRYPOINT` with the new `docker-entrypoint.sh`

For docker-compose.yml
- Remove deprecated `version`
- Remove `links` as it is not needed
- Use `redis:alpine` to reduce pull size
- Add health check to redis container
- Use `restart: unless-stopped` instead of `always` to prevent possible infinite crash-restart loop
- Adjust network name to prevent possible name collision
- Adjust app container's configuration for the new Dockerfile and docker-entrypoint.sh

For updates in app.config.ts
- Add `--db_options` arg to configure the database connection options, eg: set `tls=true` to enable tls connection

For docker-entrypoint.sh
- Configure the app with environment variable or command line arguments (all configuration options are available)
- Command line arguments have higher priority than environment variables
- No need to set `command: bash ./docker-run.sh` in `docker-compose.yml`

eg:
- `docker run -e DB_HOST=mongo NAME:TAG --db_host=localhost` will result in `node index.js --db_host=localhost`

### Additional context

`docker-entrypoint.sh` require bash 4.0+ to run. The default bash on macOS is 3.2, install bash from homebrew if you want to test it on macOS. Alpine linux (base image) won't have this problem.

Line 126 to 175 in `docker-entrypoint.sh` are the configuration options, the format is:

```bash
declare -A valueMap=(
  [ENV_VAR_NAME]="TYPE,COMMAND_LINE_FLAG,DEFAULT_VALUE"
)
```

The type could be `value` or `switch`. If the type is `value`, the command line flag should end with `=`, eg: `--port=`. If the type is `switch`, the existence of command line flag will be considered as true. The `--color` options is always true now.

The default value can be set to `@@NULL@@` which means empty. It can not be set to an empty string.

### CODE REVIEW

File: docker-entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  CMD=$@
+  CONNECTION_STRING="mongodb://"
+
+  if [ "$(is_in_cmd_with_value "--db_connection_string=" $CMD)" = "true" ]; then
+    CONNECTION_STRING=$(get_cmd_value "--db_connection_string=" $CMD)
+  else
+    if [ "$(is_in_cmd_with_value "--db_user=" $CMD)" = "true" ]; then
+      CONNECTION_STRING+="$(get_cmd_value "--db_user=" $CMD):************@"
+    fi
+    CONNECTION_STRING+="$(get_cmd_value "--db_host=" $CMD):$(get_cmd_value "--db_port=" $CMD)/$(get_cmd_value "--db_collection_name=" $CMD)"
```
Comment (ttimochan) [inline docker-entrypoint.sh]: 根据 app.config.ts ，此处应该写为 --collection_name

File: docker-entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+# ======= Environment Variables =======
+
+declare -A valueMap=(
+  [PORT]="value,--port=,2333"
+  [DEMO]="switch,--demo,false"
+  [ALLOWED_ORIGINS]="value,--allowed_origins=,localhost"
+  [CONFIG_PATH]="value,--config_path=,@@NULL@@"
+
+  # DB
+  [DB_COLLECTION_NAME]="value,--db_collection_name=,mx-space"
```
Comment (ttimochan) [inline docker-entrypoint.sh]: 同样，此处应为 --collection_name

Comment (ttimochan) [reply docker-entrypoint.sh]: 修正后，mx-server 运行正常 
image

Comment (AlisaAkiron) [reply docker-entrypoint.sh]: 文件改过好几版，改到后面忘记了，现已修复

**Review by ttimochan — APPROVED** [review]

**Review by Innei — COMMENTED** [review]
Thank you！

**Comment by Innei** [comment]
@LiamSho Docker Build CI failed, can you fix it?

**Comment by AlisaAkiron** [comment]
> @LiamSho Docker Build CI failed, can you fix it?

image
image


似乎是因为 CI Build 完成后没有 innei/mx-server:5 这个 tag （应该只有带 Git Tag 发版本才会有这个 Image Tag） 导致 Docker 又从 Docker Hub 拉了旧的 Image 下来，新的这个 Docker Image 是不会输出 MixSpace Server Image 这行 Log 的。而且这个问题好像以前一直存在（所以之前的测试都是白测

**Pull request merged**